File: C_Projects/05_Maloc/binary_tree.c
/*
 * binary_tree.c
 * By Antonio F. Huertas
 * This program defines and manipulates a binary search tree.
 */

#include <stdio.h>   // for puts, printf, scanf
#include <stdbool.h> // for bool, true, false
#include <stddef.h>  // for NULL
#include <stdlib.h>  // for malloc, free

// Defines an alias for the node structure.
typedef struct node Node;

// Declares a structure for nodes.
struct node {
    int data;
    Node *left, *right;
};

// Declares the data structure for a binary tree.
typedef struct {
    Node *root;
} Tree;

// Declares functions using prototypes.
Tree *create_tree(void);
void add_tree(Tree *bst, int val);
void show_tree(const Tree *bst);
int sum_tree(const Tree *bst);
bool search_tree(const Tree *bst, int val);
void destroy_tree(Tree *bst);

Node *add_helper(Node *curr, int val);
void show_helper(const Node *curr);
int sum_helper(const Node *curr);
bool search_helper(Node *curr, int val);
void destroy_helper(Node *curr);

// Starts the execution of the program.
int main(void) {
    Tree *bst = create_tree();
    puts("A binary search tree has been created.");

    add_tree(bst, 40);
    add_tree(bst, 10);
    add_tree(bst, 20);
    add_tree(bst, 50);
    add_tree(bst, 30);

    printf("After adding some elements, the tree is ");
    show_tree(bst);
    printf("\nThe sum of its elements is %d\n", sum_tree(bst));
    printf("\n");

    int num;
    printf("Enter a number to be searched: ");
    scanf("%d", &num);

    bool found = search_tree(bst, num);
    if (found)
        puts("The number was found!");
    else
        puts("The number was NOT found.");

    printf("\nEnter a number to be added to the tree: ");
    scanf("%d", &num);
    add_tree(bst, num);

    printf("After adding a new element, the tree is ");
    show_tree(bst);
    printf("\nThe sum of its elements is %d\n", sum_tree(bst));
    printf("\n");

    destroy_tree(bst);
    bst = NULL;
    puts("The binary search tree has been deallocated.");
    return 0;
}

// Returns a new empty tree.
Tree *create_tree(void) {
    Tree *bst = malloc(sizeof(Tree));
    bst->root = NULL;
    return bst;
}

// Adds a new value to the given tree.
void add_tree(Tree *bst, int val) {
    bst->root = add_helper(bst->root, val);
}

Node *add_helper(Node *curr, int val) {
    if(curr == NULL) {
        curr = malloc(sizeof(Node));
        curr->data = val;
        curr->left = curr->right = NULL;
    }
    else {
        if(val < curr->data)
            curr->left = add_helper(curr->left, val);
        else if(val > curr->data)
            curr->right = add_helper(curr->right, val);
        else
            curr->data = val;
    }
    return curr;
}

// Shows the contents of the given tree.
void show_tree(const Tree *bst) {
    show_helper(bst->root);
}

void show_helper(const Node *curr) {
    if(curr != NULL) {
        show_helper(curr->left);
        printf("%d ", curr->data);
        show_helper(curr->right);
    }
}

// Returns the sum of the given tree.
int sum_tree(const Tree *bst) {
    return sum_helper(bst->root);
}

int sum_helper(const Node *curr) {
    if(curr == NULL)
        return 0;
    else
        return sum_helper(curr->left) + curr->data + sum_helper(curr->right);
}

// Determines whether the given tree contains the given value.
bool search_tree(const Tree *bst, int val) {
    return search_helper(bst->root, val);
}

bool search_helper(Node *curr, int val) {
    if(curr == NULL)
        return false;
    else {
        if(val < curr->data)
            return search_helper(curr->left, val);
        else if(val > curr->data)
            return search_helper(curr->right, val);
        else
            return true;
    }
}

// Frees the memory allocated for the tree.
void destroy_tree(Tree *bst) {
    destroy_helper(bst->root);
    free(bst);
}

void destroy_helper(Node *curr) {
    if(curr != NULL) {
        destroy_helper(curr->left);
        destroy_helper(curr->right);
        free(curr);
    }
}

File: F# Workspace/05_Unions/05_Unions/binary_tree.fs
(*
 * binary_tree.fs
 * By Antonio F. Huertas
 * This program defines and manipulates a binary search tree of 
 * integers.
 *)

// Defines a discriminated union type for a binary tree.   
type IntTree =
    | Nil
    | Node of int * IntTree * IntTree

// sum_tree : IntTree -> int
// Returns the sum of the elements in the given tree.
let rec sum_tree = function
    | Nil -> 0
    | Node (root, lft, rgt) -> sum_tree lft + root + sum_tree rgt

// contains_tree : IntTree -> int -> bool
// Returns true if the given tree contains the given element.    
let rec contains_tree bst elem =
    match bst with
    | Nil -> false
    | Node (root, lft, _) 
            when elem < root -> contains_tree lft elem
    | Node (root, _, rgt) 
            when elem > root -> contains_tree rgt elem
    | _ -> true

// add_tree : IntTree -> int -> IntTree
// Returns the tree that results by adding the given element to the
// given tree.
let rec add_tree bst elem =
    match bst with
    | Nil -> Node (elem, Nil, Nil)
    | Node (root, lft, rgt) 
            when elem < root -> Node (root, (add_tree lft elem), rgt)
    | Node (root, lft, rgt) 
            when elem > root -> Node (root, lft, (add_tree rgt elem))
    | _ -> bst

// show_tree : IntTree -> unit
// Shows the elements in the given tree using inorder.
let rec show_tree = function
    | Nil -> ()
    | Node (root, lft, rgt) -> show_tree lft
                               printf "%d " root
                               show_tree rgt

// start : unit -> unit
let start () =
    let bst = Nil
    let bst1 = add_tree bst 30
    let bst2 = add_tree bst1 40
    let bst3 = add_tree bst2 10
    let bst4 = add_tree bst3 50
    let bst5 = add_tree bst4 20

    printf "This is the tree: "
    show_tree bst5
    printfn "\nThis is the tree, internally: %A" bst5
    printfn "The sum of its elements is %d" (sum_tree bst5)
    printfn "Does it contain 50? %b" (contains_tree bst5 50)